src/SharpCompress/Writers/Zip/ZipCentralDirectoryEntry.cs

```diff
@@ -13,36 +13,51 @@ internal class ZipCentralDirectoryEntry
         internal DateTime? ModificationTime { get; set; }
         internal string Comment { get; set; }
         internal uint Crc { get; set; }
-        internal uint HeaderOffset { get; set; }
-        internal uint Compressed { get; set; }
-        internal uint Decompressed { get; set; }
+        internal ulong HeaderOffset { get; set; }
+        internal ulong Compressed { get; set; }
+        internal ulong Decompressed { get; set; }
+        internal ushort Zip64HeaderOffset { get; set; }
 
         internal uint Write(Stream outputStream, ZipCompressionMethod compression)
         {
             byte[] encodedFilename = Encoding.UTF8.GetBytes(FileName);
             byte[] encodedComment = Encoding.UTF8.GetBytes(Comment);
 
-            //constant sig, then version made by, compabitility, then version to extract
-            outputStream.Write(new byte[] {80, 75, 1, 2, 0x14, 0, 0x0A, 0}, 0, 8);
+            var zip64 = Compressed >= uint.MaxValue || Decompressed >= uint.MaxValue || HeaderOffset >= uint.MaxValue || Zip64HeaderOffset != 0;
+
+            var compressedvalue = zip64 ? uint.MaxValue : (uint)Compressed;
+            var decompressedvalue = zip64 ? uint.MaxValue : (uint)Decompressed;
+            var headeroffsetvalue = zip64 ? uint.MaxValue : (uint)HeaderOffset;
+            var extralength = zip64 ? (2 + 2 + 8 + 8 + 8 + 4) : 0;
+            var version = (byte)(zip64 ? 45 : 10);
+
             HeaderFlags flags = HeaderFlags.UTF8;
             if (!outputStream.CanSeek)
             {
-                flags |= HeaderFlags.UsePostDataDescriptor;
+                // Cannot use data descriptors with zip64:
+                // https://blogs.oracle.com/xuemingshen/entry/is_zipinput_outputstream_handling_of
+                if (!zip64)
```

Comment:
This isn't true.  The post data descriptor will have 8 byte values if zip64 instead of 4 byte values,

```
 4.3.9.1 This descriptor MUST exist if bit 3 of the general
      purpose bit flag is set (see below).  It is byte aligned
      and immediately follows the last byte of compressed data.
      This descriptor SHOULD be used only when it was not possible to
      seek in the output .ZIP file, e.g., when the output .ZIP file
      was standard output or a non-seekable device.  For ZIP64(tm) format
      archives, the compressed and uncompressed sizes are 8 bytes each.
```

https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT

If there's no data descriptor then you can't have purely streaming zip creation.